### PR TITLE
implemented generic storageImpl class

### PR DESCRIPTION
**Pull request by Den-moroz**

### CODE REVIEW

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 1]
```diff
@@ -3,17 +3,52 @@
...
+        }
+        return values[index];
     }
 
     @Override
     public int size() {
+        return size;
+    }
+
+    private int contains(K key) {
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:46]: method "contains" from its name should return true or false, but this method returns position of key in your storage, please rename this method

Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:46]: What if key will be null? Is null valid value to store in storage?

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 2]
```diff
@@ -3,17 +3,52 @@
...
+    public StorageImpl() {
+        keys = (K[]) new Object[MAX_SIZE];
+        values = (V[]) new Object[MAX_SIZE];
+    }
+
     @Override
     public void put(K key, V value) {
+        int index = contains(key);
+        if (index == -1) {
+            if (size < MAX_SIZE) {
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:20]: it will be better to not make if inside of if. I would prefere to make 2 not nested if statements, first will be check for max size and if max size reached throw exception. Second one will check for index

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good, but some improvenets required

File: src/main/java/core/basesyntax/impl/StorageImpl.java (hunk 1 above)
Comment (Den-moroz) [reply src/main/java/core/basesyntax/impl/StorageImpl.java]: I tried to add a check for null, but an error occurs in the tests because null can be stored in storage

File: pom.xml [hunk 3]
```diff
@@ -25,6 +25,24 @@
             <version>4.12</version>
```
Comment (Vasul-Pankiv) [inline pom.xml:25]: why you push your pom? revert your commit

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 4]
```diff
@@ -3,17 +3,51 @@
...
+        keys = (K[]) new Object[MAX_SIZE];
+        values = (V[]) new Object[MAX_SIZE];
+    }
+
     @Override
     public void put(K key, V value) {
+        int index = indexOfKey(key);
+        if (size > MAX_SIZE) {
+            throw new RuntimeException("Storage size is full");
+        }
```
Comment (Vasul-Pankiv) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:21]: size == MAX_SIZE is invalid too
also, what if storage full, but you need to replace value for some key?
move your check to the rigth place

File: src/main/java/core/basesyntax/impl/StorageImpl.java [hunk 5]
```diff
@@ -3,17 +3,51 @@
...
+        return values[index];
     }
 
     @Override
     public int size() {
+        return size;
+    }
+
+    private int indexOfKey(K key) {
```
Comment (Vasul-Pankiv) [inline src/main/java/core/basesyntax/impl/StorageImpl.java:45]: better rename it to getIndex or getIndexByKey()

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]
